### PR TITLE
CONCF 445

### PR DESCRIPTION
**Pull request by hakubo**
Loads of changes so unit tests are failing.
But before I jump in and fix those, I want to be sure that those changes make sense.

@kvas-damian @nandy-andy @SebastianMarzjan 

I could not find how wikia.com loads images from a specific hub. Did it ever work?
More examples you can find in a ticket, here is one:
You can see all images here: http://gameshub.wikia.com/wiki/Games_Hub
but not in the APIs
http://www.wikia.com/api/v1/WikiaHubs/ModuleDataV3?module=1&city=955764

and sometimes they do:
http://www.wikia.com/api/v1/WikiaHubs/ModuleDataV3?module=5&city=955764

I made a change where I always explicitly ask GlobalFile with a correct cityId. But maybe there is something that I am missing there?

I also can not find any fallback for images made with foreignrepo.

I spent way to much time on this right now. Do you think you can help me with this one?

### CODE REVIEW

File: extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php [hunk 1]
```diff
@@ -90,8 +91,8 @@ protected function getView($viewName, $data) {
 		return $this->app->getView(get_class($this), $viewName, $data);
 	}
 
-	protected function getImageInfo($fileName, $destSize = 0) {
-		return ImagesService::getLocalFileThumbUrlAndSizes($fileName, $destSize, ImagesService::EXT_JPG);
+	protected function getImageInfo($fileName) {
+		return GlobalFile::newFromText($fileName, $this->cityId);
```
Comment (kvas-damian) [inline extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php]: Please use JPGs in output urls to improve performance.

File: extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleFromthecommunityService.class.php [hunk 2]
```diff
@@ -259,21 +260,24 @@ public function getStructuredData($data) {
...
 					'wikiUrl' => $data['wikiUrl' . $i],
 					'quote' => $data['quote' . $i],
 					'photoName' => isset($data['photo'.$i]) ? $data['photo'.$i] : ''
 				);
+
+				if( !empty($data['photo' . $i]) ) {
+					$imageData = $this->getImageInfo($data['photo' . $i]);
+
+					if ($imageData) {
+						$entries['imageAlt'] = $imageData->getName();
```
Comment (nandy-andy) [inline extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleFromthecommunityService.class.php]: I don't think this is what you want. You probably want to have `imageAlt` and `imageUrl` set for one specific entry from the `$entries`.

File: extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php (hunk 1 above)
Comment (nandy-andy) [inline extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php]: Nice. I didn't know `GlobalFile`. It probably works better than `ImagesService` which worked some time ago :smile_cat: 
And it seems it existed back when we worked on hubs services... what a pity we didn't know it.

**Comment by nandy-andy** [comment]
The code overall looks fine - just address one of my notes, please. However, I don't think I'm the right person to write "LGTM" here ;) //ping @kvas-damian

**Comment by kvas-damian** [comment]
@hakubo I double check why some API responses have URLs in place. Most likely it's because data stored in memcache was generated on a Hub wiki and right now we don't need to make those calculations again.

To verify that hypothesis I rebuilt memcache on one working API response and URLs disappeared. Then I rebuild memecahe on Hub page and API response was fixed. 

Looks that this feature never workd correctly before :(
Your solution looks ok, the only thing I'd verify is performance of ```GlobalFile```. Don't know if it's still valid but some time ago we didn't cache it enough.

File: extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php [hunk 3]
```diff
@@ -10,25 +10,36 @@
...
+		$hubWiki = WikiFactory::getWikiByID($cityId);
+		$hubDbName = $hubWiki->city_dbname;
+
+		$wgForeignFileRepos[] = [
+			'class'            => 'WikiaForeignDBViaLBRepo',
+			'name'             => $hubDbName,
+			'transformVia404'  => true,
+			'wiki'             => $hubDbName,
+			'backend'          => "wikia{$hubDbName}-backend"
+		];
```
Comment (kvas-damian) [inline extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php]: Did you check what happen if you add foreign repo to the wiki that you're currently on? eg. render hub page

File: extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php [hunk 4]
```diff
@@ -269,6 +269,9 @@ protected function getDefaultModuleList() {
 		return $out;
 	}
 
+	/**
+	 * @return User
```
Comment (rogatty) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: This is a string: `protected $userClass = 'User';`

File: extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php [hunk 5]
```diff
@@ -299,6 +302,9 @@ public function getModuleUrl( $timestamp, $moduleId ) {
 		);
 	}
 
+	/**
+	 * @return SpecialPage
```
Comment (rogatty) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: This is a string too: `protected $specialPageClass = 'SpecialPage';`

File: extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php [hunk 6]
```diff
@@ -551,9 +557,11 @@ public function getCorporateWikisLanguages() {
 	public function getVideoData ($fileName, $thumbSize) {
 		$videoData = [];
 		$title = Title::newFromText($fileName, NS_FILE);
+
 		if (!empty($title)) {
-			$file = wffindFile($title);
+			$file = wfFindFile($title);
```
Comment (rogatty) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: Wow... What was happening when it got here?

File: extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php (hunk 3 above)
Comment (hakubo) [inline extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleService.class.php]: It does not get there, as it'll find the file in local repo.

File: extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php (hunk 6 above)
Comment (hakubo) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: it... just... worked...

File: extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php (hunk 5 above)
Comment (hakubo) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: but on like 295 it is used as SpecialPage Class. wtf?

Comment (rogatty) [inline extensions/wikia/WikiaHubsServices/models/EditHubModel.class.php]: http://stackoverflow.com/a/3121559/1050577

It's basically the same as `new $className()`

**Comment by rogatty** [comment]
Fix types in comments, fix tests and it LGTM.

**Comment by rogatty** [comment]
Tests are failing with:
```
Running 'HubRssModelTest'.....
Strict Standards: Declaration of WikiaHubsModuleSliderService::getImageInfo() should be compatible with WikiaHubsModuleService::getImageInfo($fileName, $destSize = 0) in /var/lib/jenkins/workspace/unitTests_php_branches/extensions/wikia/WikiaHubsServices/modules/WikiaHubsModuleSliderService.class.php on line 2
.
Catchable fatal error: Argument 1 passed to Mock_WikiaHubsModuleSliderService_4fc7e173::loadData() must be an instance of EditHubModel, null given, called in /var/lib/jenkins/workspace/unitTests_php_branches/extensions/wikia/HubRssFeed/models/external/HubRssFeedModel.class.php on line 127 and defined in /var/lib/jenkins/workspace/unitTests_php_branches/lib/composer/phpunit/phpunit-mock-objects/src/Framework/MockObject/Generator.php(313) : eval()'d code on line 11
```

so merging dev won't help :P

**Comment by rogatty** [comment]
LGTM

**Comment by bognix** [comment]
LGTM, tested, merging

**Pull request merged**